### ISSUE TITLE
discard lines not working 

### ISSUE THREAD

**Issue by Justin-JHG** [desktop/desktop#10083]
### Describe the bug

Discard lines not working if changes involving delete lines

### Version & OS

2.5.3 on windows 10

### Steps to reproduce the behavior

1. Go to any file in the repo and remove one line then add 2 lines
2. Click on any modified lines above and right click 

### Expected behavior

Discard modified line appears and enabled 

### Actual behavior

Discard modified line appears but disabled

### Screenshots
![image](https://user-images.githubusercontent.com/41654878/85652730-dcbbfa80-b6ee-11ea-8924-352608a2eaab.png)

**Comment by billygriffin** [comment]
Hi @Justin-JHG, thanks for the issue! This was an intentional design choice that we made to avoid confusion due to line numbers clashing and uncertainty around what you're actually discarding. When there are both added and deleted lines in the same hunk, in testing it out it was really tricky to figure out what would be discarded and what the state of the diff would be after the discard operation was complete. So we decided to make it disabled in that one case.

We have discussed potentially examining whether an "Undo" for this operation would allow a bit more flexibility while not risking unintended data loss. Since this just shipped, I'm going to keep this issue open to collect more feedback from others and see if it's a case that people are frequently running into, and if so we'll evaluate the path forward.

cc: @ampinsk @rafeca

**Comment by Justin-JHG** [comment]
thank you @billygriffin @ampinsk @rafeca for your hard work.
yes this is a tricky one but very useful to have, I did encounter cases in which I only add new lines and still can't discard.
the line before the new lines was somehow treated as removed line(as in red colour display) so not sure whether that's an issue with Git or Desktop.

hopefully there will be some good ideas come out to make this work. thank you.
see below example 
![image](https://user-images.githubusercontent.com/41654878/85800779-ea23c400-b784-11ea-8e50-0a06d822f73e.png)

**Comment by rafeca** [comment]
Hey @Justin-JHG ! Thanks again for reporting this issue

> the line before the new lines was somehow treated as removed line(as in red colour display) so not sure whether that's an issue with Git or Desktop.

This is caused by `git` diff behaviour: since there's no final newline at the end of the file, when adding new lines to the file the previously last line gets modified (since a newline is added there). This can be usually solved by always adding newlines at the end of files. Most editors have this option, like VSCode:

<img width="549" alt="Screen Shot 2020-06-26 at 10 50 05" src="https://user-images.githubusercontent.com/408035/85839186-e864fb00-b79a-11ea-9ed1-d827fc896412.png">

**Comment by jabamaus** [comment]
Hi,

I’ve been frustrated by this. I’ve been able to discard an added line, but never a modified one. Just feels like a broken feature to the end user if I’m honest. I’ll keep an eye on it but it definitely feels like there is room for improvement. I respect the complexities underneath though 😀

Cheers

**Comment by Justin-JHG** [comment]
> Hi,
> 
> I’ve been frustrated by this. I’ve been able to discard an added line, but never a modified one. Just feels like a broken feature to the end user if I’m honest. I’ll keep an eye on it but it definitely feels like there is room for improvement. I respect the complexities underneath though 😀
> 
> Cheers

yes same, I have not been able to discard a single line since this feature come out. Feels very incomplete

**Comment by Justin-JHG** [comment]
> Hi,
> 
> I’ve been frustrated by this. I’ve been able to discard an added line, but never a modified one. Just feels like a broken feature to the end user if I’m honest. I’ll keep an eye on it but it definitely feels like there is room for improvement. I respect the complexities underneath though 😀
> 
> Cheers

Hi @jabamaus 
I just found out to discard modified lines you need to click the area between the line number and the +/- sign. hope this helps.
https://github.com/desktop/desktop/issues/10194#issuecomment-657617972

**Comment by jabamaus** [comment]
Thanks Justin! Appreciate that.

**Comment by adamreisnz** [comment]
That seems like a bug, clicking on the line number shows the discard option, but slightly to the left of it doesn't. It all looks like one clickable area to me.

![image](https://user-images.githubusercontent.com/490562/97796073-85b4cc00-1c72-11eb-9ec4-67f241e85e88.png)

![Uploading image.png…]()

**Comment by Justin-JHG** [comment]
yes @adamreisnz it requires to click the highlighted line below to get it work. 
![image](https://user-images.githubusercontent.com/41654878/98050998-06e4ac80-1e87-11eb-8703-024c958f7de6.png)

I guess the reason was you can only discard multiple lines as a whole, i.e. a logic group. when clicking line number Desktop can't make sure the whole group of changes are selected to be discard

**Comment by sgreifeneder** [comment]
Justin is absolute right, thanks for the tip. This is definitely bad UX and needs to be fixed.

**Comment by LogicNg** [comment]
Any progress?

**Comment by SalmanTahir7** [comment]
This is still an issue

**Comment by DanielRouxSA** [comment]
This is definitely either a bug or a design oversight. It would be good to see this fixed.

**Comment by DanielRouxSA** [comment]
Having looked at the explanation in the comment linked by https://github.com/desktop/desktop/issues/10083#issuecomment-658493523, I understand that this appears to be an intentional design decision, and I see now how it works; hovering over the gutter highlights the whole chunk as opposed to one of the individual lines, which then allows you to revert the chunk.

However, I feel like there is potential room for improvement to the UX, considering the existence of this issue and how many people have responded either via comment or reaction.

I think the core of the issue is that when a user goes to revert an individual line change, they likely see each changed line as an individual change, but when they go to revert a modified line or lines, they intuitively see the chunk of - and + lines as a single change, and so the intuitive action to take is to right-click anywhere in the chunk. On top of this, the gutter to the right - at least to me - does not stand out has potentially having different interaction behavior than the line numbers.

I do however acknowledge that this is not a simple fix to make, because simply allowing the user to discard a chunk by right-clicking anywhere in the line numbers area would be problematic in scenarios such as a line being modified, and then a new, potentially totally unrelated line being added right underneath it - because then how does one go about, for example, letting the user revert just the changed line without touching the added one?

This will definitely require some thinking to improve.

**Comment by Justin-JHG** [comment]
agree with you @DanielRouxSA 
also I think there are some imporvements required in this area, maybe when hover the mouse then a button would pop up to allow you to discard changes, instead of selecting the line/lines(which took time to position the mouse cursor to there) and then need to right click and then select discard. took me ages everytime to discard the changes I don't want to commit

**Comment by SpicyCatGames** [comment]
Once you select a line with line number, if a line is so long that it's shown as two or more lines, all are selected as a group. There could be a button on top that will be greyed out until lines are selected, showing "discard changes on selected lines". I feel that it would be much more intuitive compared to the current behavior.

**Comment by calebrader** [comment]
Please make the "highlighted area" wider. It's not obvious to the user that they need to click this narrow region to discard modified lines. Thank you Justin for pointing this out.

**Comment by jdimeo** [comment]
I've used VS Code, SourceTree, GitKraken, and Fork and this _totally_ threw me for a loop- I usually can figure this kind of stuff out

**Comment by Justin-JHG** [comment]
A hidden gem 💎

Sent from my iPhone
________________________________

**Comment by abrambailey** [comment]
I too was very confused by this behavior. I would suggest putting up an error message when someone tries to click discard modified line, to explain this situation, rather than just greying it out.

**Comment by SpicyCatGames** [comment]
> I too was very confused by this behavior. I would suggest putting up an error message when someone tries to click discard modified line, to explain this situation, rather than just greying it out.

Why? Because a feature is risky, instead of warning the user, let's make it nearly unusable. That's what I got from this thread.

**Comment by HughWarrington** [comment]
> Please make the "highlighted area" wider. It's not obvious to the user that they need to click this narrow region to discard modified lines. Thank you Justin for pointing this out.

Yeah, that margin is pretty ludicrously thin. Very hard for me to hit using the mouse as an able-bodied person. https://craftcms.com/blog/accessible-target-sizes

**Comment by HughWarrington** [comment]
I see in a recent update that this narrow gutter which you must hit in order to bring up the desired context menu has been moved to the left of the line numbers, and the colour contrast reduced.

Before:
![image](https://github.com/desktop/desktop/assets/5960770/97ae1d70-9d60-4609-b609-faf18089fc3e)

After:
![image](https://github.com/desktop/desktop/assets/5960770/e55f157f-d5f2-4318-a00d-27dab33ffdd3)

Also the gutter is now immediately next to the column of pixels which activate the horizontal resize handle between the two halves of this form. This makes the context menu even less discoverable.

**Comment by SpicyCatGames** [comment]
If a feature is risky, also make it unintuitive🤔
